blsqpy/extract.py

```diff
@@ -0,0 +1,32 @@
+import numpy as np
+
+from blsqpy.descriptor import Descriptor
+
+
+def to_data_elements(activity):
+    data_elements = []
+    for _state_code, state in Descriptor.as_items(activity.states):
+        for _source_code, source in Descriptor.as_items(state.sources):
+            data_elements.extend(source.uids)
+    return data_elements
+
+
+def rotate_de_coc_as_columns(df):
+    df = df.rename(index=str, columns={
+        "uidorgunit": "orgunit",
+        "catcomboid": "coc",
+        "dataelementid": "de"
+    })
+
+    df["period"] = np.where(df.monthly, df.monthly, df.quarterly)
+    df = df.drop(["uidlevel3", "uidlevel2", "enddate",
```

Comment:
We want to keep "uidlevel3", "uidlevel2" in at that point as we use them for aggregation.

Also, depending on extracts / countries, the relevant grouping will be a different level. would need to paramatrize that accross extraction and mapping.

Reply:
I think ou will be part of another data frame that you will join later (at aggregation)